packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts

```
@@ -0,0 +1,73 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { PathAccess } from '../types/credentials';
+import { BucketInfo } from '../../providers/s3/types/options';
+import { ENTITY_IDENTITY_URL } from '../utils/constants';
+import { StorageAccess } from '../types/common';
+
+const resolvePermissions = (
+	accessRule: Record<string, string[]>,
+	token: boolean,
+	groups?: string,
+) => {
+	if (!token) {
+		return {
+			permission: accessRule.guest,
+		};
+	}
+	if (groups) {
+		const selectedKey = Object.keys(accessRule).find(
+			access => access.includes(groups) || access.includes('authenticated'),
+		);
+
+		return {
+			permission: selectedKey ? accessRule[selectedKey] : undefined,
+		};
+	}
+
+	return {
+		permission: accessRule.authenticated,
+	};
+};
+
+export const generateLocationsFromPaths = ({
+	buckets,
+	tokens,
+	identityId,
+	userGroup,
+}: {
+	buckets: Record<string, BucketInfo>;
+	tokens: boolean;
+	identityId?: string;
+	userGroup?: string;
+}): PathAccess[] => {
+	const locations: PathAccess[] = [];
+
+	for (const [, bucketInfo] of Object.entries(buckets)) {
+		const { bucketName, paths } = bucketInfo;
+		if (!paths) {
+			// Todo: Verify behavior
+			return locations;
+		}
+		for (const [path, accessRules] of Object.entries(paths)) {
+			if (tokens && identityId && path.includes(ENTITY_IDENTITY_URL)) {
+				locations.push({
+					type: 'PREFIX',
+					permission: accessRules.entityidentity as StorageAccess[],
+					scope: {
+						bucketName,
+						path: path.replace(ENTITY_IDENTITY_URL, identityId),
+					},
+				});
+			}
+			const location = {
+				type: 'PREFIX',
+				...resolvePermissions(accessRules, tokens, userGroup),
+				scope: { bucketName, path },
+			};
+			if (location.permission) locations.push(location as PathAccess);
```

Sanity check, what are the cases when a path is omitted from `locations`? I'm wondering if we should fail-fast in these scenarios.

---

Nit: Looking at our doc it seems backend also support `Functions` access type. It make sense to not support it here, but just make sure we document it.

---

> what are the cases when a path is omitted from locations?
- There is logic in `resolvePermissions` on what permission to pick for what use cases like authenticated, unAuthenticated, user group, entity etc. Whereever we could fail fast that is done. This logic here is more santizing the array after resolvePermission.

---

> Nit: Looking at our doc it seems backend also support Functions access type. It make sense to not support it here, but just make sure we document it.

This will result in empty `paths` it actually would not even make it to `buckets` but for both of these scenarios we fail fast so it would not even get to this point